mlflow/langchain/_langchain_autolog.py

```diff
@@ -180,6 +228,682 @@ def _chain_with_retriever(model):
+
+        def _end_span(
+            self,
+            span: MLflowSpanWrapper,
+            outputs=None,
+            attributes=None,
+            status=SpanStatus(TraceStatus.OK, ""),
+        ):
+            """Close MLflow Span (or Trace if it is root component)"""
+            self._mlflow_client.end_span(
+                request_id=span.request_id,
+                span_id=span.span_id,
+                outputs=outputs,
+                attributes=attributes,
+                status=status,
+            )
+
+        def _reset(self):
+            self._run_span_mapping = {}
+
+        def _log_artifact(self, data: Dict[str, Any], filename: str):
+            """Log an artifact to the run."""
+            mlflow.log_dict(
+                data, os.path.join(self.artifacts_dir, f"{filename}.json"), run_id=self.run_id
+            )
+
+        def _analyze_text(self, text: str):
+            try:
+                import textstat
+            except ImportError:
+                warnings.warn(
+                    "Please install `textstat` for generating text complexity metrics "
+                    "for the text."
+                )
+                return
+            else:
+                return {key: getattr(textstat, key)(text) for key in TEXT_COMPLEXITY_METRICS}
+
+        @staticmethod
+        def _get_stacktrace(error: BaseException) -> str:
+            """Get the stacktrace of the parent error."""
+            msg = repr(error)
+            try:
+                if sys.version_info < (3, 10):
+                    tb = traceback.format_exception(error.__class__, error, error.__traceback__)
+                else:
+                    tb = traceback.format_exception(error)
+                return (msg + "\n\n".join(tb)).strip()
+            except Exception:
+                return msg
+
+        @override
+        def on_chat_model_start(
+            self,
+            serialized: Dict[str, Any],
+            messages: List[List[BaseMessage]],
+            *,
+            run_id: UUID,
+            tags: Optional[List[str]] = None,
+            parent_run_id: Optional[UUID] = None,
+            metadata: Optional[Dict[str, Any]] = None,
+            name: Optional[str] = None,
+            **kwargs: Any,
+        ):
+            """Run when a chat model starts running."""
+            if metadata:
+                kwargs.update({"metadata": metadata})
+            llm_inputs = {"messages": [[dumpd(msg) for msg in batch] for batch in messages]}
+            llm_span = self._start_span(
+                span_name=name or "chat model",
+                parent_run_id=parent_run_id,
+                # we use LLM for chat models as well
+                span_type=SpanType.LLM,
+                run_id=run_id,
+                inputs=llm_inputs,
+                attributes=kwargs,
+            )
+            llm_span.add_event(
+                SpanEvent("start", timestamp=int(datetime.now(timezone.utc).timestamp()))
```

Comment:
What is the purpose of this event? A span already contains start timestamp.

Reply:
Langchain logs this event, it might be useless in most of the cases. but when something breaks do we always end the span and emit traces? In that case SpanEvents will only have `start` event but no `end` event, which could be informative

Reply:
Such span will have null end_time, would it be sufficient to tell the span is not ended?